### PR TITLE
Feature: DISTS metric

### LINKED ISSUE #123: Metric: DISTS
**Is your feature request related to a problem? Please describe.**
DISTS is a novel method of measuring image quality based on VGG16 architecture.
Python code: https://github.com/dingkeyan93/DISTS/blob/master/DISTS_pytorch/DISTS_pt.py
Original article: https://arxiv.org/abs/2004.07728

**Describe the solution you'd like**
Copy authors implementation with code refactoring. 
It's easy to inherit existing `ContentLoss` class and build DISTS on top of it

### PR DESCRIPTION
**Pull request by zakajd**
Closes #123 

## Proposed Changes

  - DISTS metric
 - Renamed `use_average_pooling` to `replace_pooling` because in DISTS pooling is replaced by L2pooling layer, not AveragePool. 
  - Moved `__init__` descriptions to class description, so that it's visible in VS Code and Jupyter IDA
  - Reduced image size in test to make them faster
  - Fixed small bug in FSIM testing

### CODE REVIEW

File: tests/test_perseptual.py [hunk 1]
```diff
@@ -128,10 +129,69 @@ def test_lpips_loss_forward(input_tensors: Tuple[torch.Tensor, torch.Tensor], de
...
+def test_dists_loss_backward(prediction: torch.Tensor, target: torch.Tensor, device: str) -> None:
+    loss = LPIPS()
+    loss(prediction.to(device), target.to(device))
+
+
+@pytest.mark.parametrize(
+    "prediction,target,expectation,value",
+    [
+        (torch.zeros(4, 3, 128, 128), torch.zeros(4, 3, 128, 128), raise_nothing(), 0.0),
+        (torch.ones(4, 3, 128, 128), torch.ones(4, 3, 128, 128), raise_nothing(), 0.0),
```
Comment (snk4tr) [inline tests/test_perseptual.py:161]: I assume you wanted to reduce resolution here to 96 as well

File: tests/test_perseptual.py [hunk 2]
```diff
@@ -71,10 +72,10 @@ def test_content_loss_supports_custom_extractor(prediction: torch.Tensor, target
...
     "prediction,target,expectation,value",
     [
-        (torch.rand(4, 3, 128, 128, 2), torch.rand(4, 3, 128, 128, 2), pytest.raises(AssertionError), None),
-        (torch.randn(4, 3, 128, 128), torch.randn(4, 3, 128, 128), pytest.raises(AssertionError), None),
-        (torch.zeros(4, 3, 128, 128), torch.zeros(4, 3, 128, 128), raise_nothing(), 0.0),
-        (torch.ones(4, 3, 128, 128), torch.ones(4, 3, 128, 128), raise_nothing(), 0.0),
+        (torch.rand(4, 3, 96, 96, 2), torch.rand(4, 3, 96, 96, 2), pytest.raises(AssertionError), None),
+        (torch.randn(4, 3, 96, 96), torch.randn(4, 3, 96, 96), pytest.raises(AssertionError), None),
+        (torch.zeros(4, 3, 96, 96), torch.zeros(4, 3, 96, 96), raise_nothing(), 0.0),
+        (torch.ones(4, 3, 96, 96), torch.ones(4, 3, 96, 96), raise_nothing(), 0.0),
```
Comment (snk4tr) [inline tests/test_perseptual.py:78]: I think you can also reduce number of elements in the batch here to 2 without the loss of generality

File: tests/test_perseptual.py [hunk 3]
```diff
@@ -128,10 +129,69 @@ def test_lpips_loss_forward(input_tensors: Tuple[torch.Tensor, torch.Tensor], de
     loss(prediction.to(device), target.to(device))
 
 
+def test_lpips_loss_backward(prediction: torch.Tensor, target: torch.Tensor, device: str) -> None:
+    loss = LPIPS()
+    loss(prediction.to(device), target.to(device))
```
[the remark this answers is not in the document]
Comment (zakajd) [reply tests/test_perseptual.py:134]: Done. I deleted separate test for backward and now only test that gradients are computed correctly.

File: tests/test_perseptual.py [hunk 4]
```diff
@@ -128,10 +129,69 @@ def test_lpips_loss_forward(input_tensors: Tuple[torch.Tensor, torch.Tensor], de
...
+@pytest.mark.parametrize(
+    "prediction,target,expectation,value",
+    [
+        (torch.zeros(4, 3, 128, 128), torch.zeros(4, 3, 128, 128), raise_nothing(), 0.0),
+        (torch.ones(4, 3, 128, 128), torch.ones(4, 3, 128, 128), raise_nothing(), 0.0),
+    ],
+)
+def test_dists_loss_forward_for_special_cases(
+        prediction: torch.Tensor, target: torch.Tensor, expectation: Any, value: float) -> None:
+    loss = ContentLoss()
```
[the remark this answers is not in the document]
Comment (zakajd) [reply tests/test_perseptual.py:166]: Left from test copy-paste. Fixed now.

File: piq/perceptual.py [hunk 5]
```diff
@@ -251,38 +251,142 @@ class LPIPS(ContentLoss):
...
         super().__init__("vgg16", layers=lpips_layers, weights=lpips_weights,
-                         use_average_pooling=use_average_pooling, distance=distance,
+                         replace_pooling=replace_pooling, distance=distance,
                          reduction=reduction, mean=mean, std=std,
                          normalize_features=True)
+
+
+class DISTS(ContentLoss):
+    r"""Deep Image Structure and Texture Similarity metric.
+    Expects input to be in range [0, 1] or normalized with ImageNet statistics into range [-1, 1]
```
[the remark this answers is not in the document]
Comment (zakajd) [reply piq/perceptual.py:291]: Yes, but `forward` pass fails because of `_validate_input` function :\

File: piq/perceptual.py [hunk 6]
```diff
@@ -251,38 +251,142 @@ class LPIPS(ContentLoss):
...
+            If there is no need to normalize data, use [1., 1., 1.].
+
+    References:
+        .. [1] Keyan Ding, Kede Ma, Shiqi Wang, Eero P. Simoncelli
+        (2020). Image Quality Assessment: Unifying Structure and Texture Similarity.
+        https://arxiv.org/abs/2004.07728
+        .. [2] https://github.com/dingkeyan93/DISTS
+    """
+    # Constant used in feature normalization to avoid zero division
+    EPS = 1e-10
```
[the remark this answers is not in the document]
Comment (zakajd) [reply piq/perceptual.py:308]: Agree, moved it to the top

File: tests/test_perseptual.py (hunk 1 above)
Comment (zakajd) [reply tests/test_perseptual.py:161]: Yes, missed this lines

File: tests/test_perseptual.py (hunk 2 above)
Comment (zakajd) [reply tests/test_perseptual.py:78]: Done

File: piq/perceptual.py [hunk 7]
```diff
@@ -251,38 +251,142 @@ class LPIPS(ContentLoss):
...
+    def construct_kernel(self, x: torch.Tensor) -> torch.Tensor:
+        r"""Returns 2D Hann window kernel with number of channels equal to input channels"""
+        C = x.size(1)
+        
+        # Take bigger window and drop borders
+        window = torch.hann_window(self.kernel_size + 2, periodic=False)[1:-1]
+        kernel = window[:, None] * window[None, :]
+
+        # Normalize and reshape kernel
+        self.kernel = (kernel / kernel.sum()).repeat((C, 1, 1, 1)).to(x)
```
[the remark this answers is not in the document]
Comment (zakajd) [reply piq/perceptual.py]: I did this refactoring in a single commit. It's not so big to require a separate PR

**Comment by zakajd** [comment]
Also closed #132  in one of the commits

**Comment by zakajd** [comment]
@snk4tr Ready to be merged
